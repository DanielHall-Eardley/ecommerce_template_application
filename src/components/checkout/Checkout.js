import React, {useState} from 'react'

import {connect} from 'react-redux'
import {
  displayError, 
  clearError,
  displayNotification,
  clearNotification
} from '../../actions/notification'
import {storeOrder, clearOrder} from '../../actions/order'
import {storeUser, clearUser} from '../../actions/user'

import detectCssSupport from '../../helper/detectCssSupport'
import Loading from './Loading'

import PostageContainer from './PostageContainer'
import ProductSummary from './ProductSummary'
import CheckoutSummary from './CheckoutSummary'
import PaymentForm from './PaymentForm'
import Address from './Address'

import defaultStyles from './Checkout.module.css'
import fallBackStyles from './CheckoutFallback.module.css'
let styles = defaultStyles

if (!detectCssSupport()) {
  styles = fallBackStyles
}

export const Checkout = props => {
  const [clientSecret, setSecret] = useState(null)
  const [loading, setLoading] = useState(false)
  const [loadingMsg, setLoadingMsg] = useState('')

  
  /*This function makes a request to the api to retrieve the current order*/
  const getCheckoutSummary = async (userId, token) => {
    const url  = `/checkout/summary/${props.orderId}/${userId}`
    const response = await props.getAuthApi(url, token, props.displayError)

    if (response) {
      props.storeOrder(response.order)
      setSecret(response.clientSecret)
    }
  }

  
  /*This function checks if a user is logged in. 
  If a current order exists and the user is of type: customer,
  the getOrderSummary function is called*/
  React.useEffect(() => {
    props.clearError()
    
    const result = props.checkLogin()

    if (result.error) {
      props.clearUser()
      props.displayError(result.error)
    } else {
      props.storeUser(result.user)

      if(result.user.type === 'customer' && props.orderId) {
        getCheckoutSummary(result.user.userId, result.user.token)
      }
    } 
  }, [props.orderId])

  const {order} = props

  return (
    <main aria-busy={loading}>
      { loading ? <Loading msg={loadingMsg}/> : null}
      { order ? 
        <div className={styles.checkout} key={order._id} aria-label='checkout'>
          <h3 className={styles.header}>
            <span>Order Id: {order._id}</span>
          </h3>
          <ProductSummary authApi={props.authApi}/>
          <h3 className={styles.header}>Delivery Address</h3>
          { !order.addressConfirmed ? 
          <Address 
            setLoading={setLoading} 
            loading={loading}
            setLoadingMsg={setLoadingMsg}
            authApi={props.authApi}/>
          : <p className={styles.addressConfirmed}>Address Confirmed</p> }
          <h3 className={styles.header}>Postage</h3>
          <PostageContainer   
            setLoading={setLoading} 
            loading={loading}
            setLoadingMsg={setLoadingMsg}
            authApi={props.authApi}
            setSecret={setSecret}/>
          <h3 className={styles.header}>Summary</h3>
          <CheckoutSummary order={order}/>
          <h3 className={styles.header}>Complete Payment</h3>
          <PaymentForm
            authApi={props.authApi}
            clientSecret={clientSecret} 
            loading={loading}
            setLoading={setLoading}
            setLoadingMsg={setLoadingMsg}/>
        </div> 
      : <p className={styles.fallBack}>Add items to cart to create order</p>}
    </main>
  )
}

const mapStateToProps = state => {
  return {
    order: state.order.order,
    user: state.user,
    orderId: state.order.summary.orderId
  }
}

const mapDispatchToProps = dispatch => {
  return {
    displayError: error => dispatch(displayError(error)),
    storeOrder: order => dispatch(storeOrder(order)),
    storeUser: order => dispatch(storeUser(order)),
    clearError: () => dispatch(clearError()),
    clearUser: () => dispatch(clearUser()),
    clearOrder: () => dispatch(clearOrder()),
    displayNotification: notfication => dispatch(displayNotification(notfication)),
    clearNotification: () => dispatch(clearNotification()),
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Checkout)

