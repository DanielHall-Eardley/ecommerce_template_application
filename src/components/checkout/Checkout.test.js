import React from 'react'; 

import {shallow} from 'enzyme'
import {Checkout} from './Checkout';

let props 

beforeEach(() => {
  props = {
    getAuthApi: jest.fn(),
    authApi: jest.fn(),
    clearError: jest.fn(),
    displayError: jest.fn(),
    clearUser: jest.fn(),
    storeUser: jest.fn(),
    storeOrder: jest.fn(),
    clearOrder: jest.fn(),
    displayNotification: jest.fn(),
    clearNotification: jest.fn(),
    order: {
      customerId: 'id',
      customerName: 'name',
      customerEmail: 'email',
      products: [],
      total: 100,
      count: 1
    },
    user: {
      token: 'token',
      userId: 'userId',
      name: 'name',
      type: 'customer',
      tokenExpiration: 'tokenExpiration'
    },
    orderId: 'orderId'
  }

  props.checkLogin = jest.fn().mockReturnValue({user: props.user})
})  

const useEffect = jest.spyOn(React, 'useEffect').mockImplementation(f => f());

test('test Checkout snapshot', () => {
  const wrapper = shallow(<Checkout {...props}/>)
  expect(wrapper).toMatchSnapshot()
})




