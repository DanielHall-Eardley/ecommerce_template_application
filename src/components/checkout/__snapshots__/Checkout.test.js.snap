// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Checkout snapshot 1`] = `ShallowWrapper {}`;
